doc/sphinx/proof-engine/ltac.rst

```diff
@@ -52,127 +66,98 @@ the variable :token:`ident` matches any complex expression with (possible)
 dependencies in the variables :n:`@ident__i` and returns a functional term
 of the form :n:`fun @ident__1 … ident__n => @term`.
 
-The main entry of the grammar is :n:`@ltac_expr`. This language is used in proof
-mode but it can also be used in toplevel definitions as shown below.
+.. note::
+
+   - If you want :n:`@tactic__2; @tactic__3` to be fully applied to the first
+     subgoal generated by :n:`@tactic__1` before applying it to the other
+     subgoals, then you should write:
+
+     - :n:`@tactic__1; [> @tactic__2; @tactic__3 .. ]` rather than
+
+     - :n:`@tactic__1; (@tactic__2; @tactic__3)`.
+
+   .. todo PR I tried to improve the wording for the following bullet,
+      but I'm not quite sure what it refers to.  Perhaps a fool's errand.
+      Does this issue still come up, and if so, exactly which productions?
+
+   - In :token:`tactic_arg_compat`, there is a grammatical ambiguity in
+     how to interpret :token:`qualid`.  It could be either a tactic argument
+     or just a :token:`term`.  This is resolved by checking first whether there
+     is a tactic by that name, then whether it is a reference to a term and
+     using the first one that's found.
+
+     The default interpretation can be overridden by using :n:`ltac:(@qualid)` or
+     :n:`(@qualid)` to force interpretation as, respectively, a tactic or a term.
+
+     .. todo PR With the described behavior, it would never be necessary to use
+        the "ltac:(@qualid)" form since it's checked first.
+
+.. insertprodn ltac_expr tactic_atom
+
+.. prodn::
+   ltac_expr ::= {| @ltac_expr4 | @binder_tactic }
+   let_clause ::= @name := @ltac_expr
+   | @ident {+ @name } := @ltac_expr
+   ltac_expr4 ::= @ltac_expr3 ; {| @ltac_expr3 | @binder_tactic }
+   | @ltac_expr3 ; [ {? @multi_goal_tactics } ]
+   | @ltac_expr3 ; [> {? @multi_goal_tactics } ]
+   | @ltac_expr3
+   multi_goal_tactics ::= {? @ltac_expr } %| @multi_goal_tactics
+   | {? @ltac_expr } .. {? %| {*| {? @ltac_expr } } }
+   | @ltac_expr
```

Comment:
I spent way too much time on this one.  It would be much simpler if there wasn't a `|` after `..`.  I prefer 3b because it clearly breaks out the ltac_expr before `..`, which the text needs to refer to, whereas with yours I would have to use phrasing such as "The last ltac_expr before the `..`" and "All but the last ltac_expr before the `..`".  I could also go for 3.  Avoiding the separator character in the lists helps.

![image](https://user-images.githubusercontent.com/1253341/81488533-dba63b00-921e-11ea-82bb-a2b76b2b830f.png)

This is from fullGrammar.  Note the empty production.

```
tactic_then_gen: [   (* multi_goal_tactics *)
| tactic_expr5 "|" tactic_then_gen
| tactic_expr5 ".." tactic_then_last
| ".." tactic_then_last
| tactic_expr5
| "|" tactic_then_gen
|
]

tactic_then_last: [
| "|" LIST0 ( OPT tactic_expr5 ) SEP "|"
|
]
```

As for a name, how about "for_each_goal" or "apply_for_each"?

Reply:
I understand your concern with my proposal but I still find yours too hard to read. Here is an easy to read proposal inspired by your `3b` version:

```
multi_goal_tactics ::= @tactic_list
| {? @tactic_list | } {? @ltac_expr } .. {? | @tactic_list }
```

I think it shows clearly that there are two cases, one base case and one with a central `..`
`tactic_list` is unchanged wrt the previous proposal.

`for_each_goal` sounds like a reasonable name to me.

Reply:
OK, that seems good:

![image](https://user-images.githubusercontent.com/1253341/81510213-0b0f8300-92c5-11ea-819d-473ddc258f62.png)

How about `for_each_tactics` or `for_each_tacs` instead of `tactic_list`?

Reply:
Since this is repeated several times, this should be kept rather small. The current version does look OK to me.